Fixed wrong is_air check

J-co opened the pull request:
If you were to add a dielectric loss to the air gaps eps==1 -> false and it would assume propagation inside the disks, meaning no mode mixing!
Fortunately, as far as I can tell, all published results with transformer didn't include losses. Losses were so far only treated in 1D. 
I changed the check to real(eps)==1 for now. In the future one could define the propagation model for each gap if for example we want to simulate "air gaps" with slighlty different epsilon (e.g. if we have residual helium inside)...

J-co committed:
Fixed wrong is_air check

If you were to add a dielectric loss to the air gaps eps==1 -> false and it would assume propagation inside the disks, meaning no mode mixing!
Fortunately, as far as I can tell, all published results with transformer didn't include losses. Losses were so far only treated in 1D. 
I changed the check to real(eps)==1 for now. In the future one could define the propagation model for each gap if for example we want to simulate "air gaps" with slighlty different epsilon (e.g. if we have residual helium inside)...

J-co commented:
Actually,
this uncovered an inconsistency between how losses are treated in 1D vs 3d air vs 3d dielectric:
For exp(-iwt) convention a loss comes from a **negative** imaginary permittivity (eps  = eps*(1-tanD)).

In the 1D propagator (propagator1D in boostfractor.jl) we have:
```
  k0 = 2*pi/lambda
  k_prop = conj(sqrt.(k0^2))
  e1 = E0.*exp(-1im*k_prop*dz*sqrt(eps))
```
which leads to loss for negative Im(eps) since sqrt(eps) conserves the sign of the imaginary part of eps.


For 3D disk propagation (propagation_matrix(...) in boostfractor_transformer.jl) we have:
`    k0 = 2pi/lambda*sqrt(eps)`

and then:
```
kz = sqrt(k0^2 - modes.mode_kt[m,l+modes.L+1]^2)
propagation_matrix[(m-1)*(2modes.L+1)+l+modes.L+1, (m-1)*(2modes.L+1)+l+modes.L+1] = exp(-1im*kz*dz)
```
which also leads to loss for negative Im(eps)

and then finally for 3D propagation in air (propagatorNoTilts in boostfractor.jl) we have:

```
k0 = 2*pi/lambda*sqrt(eps)
k_prop = [conj(sqrt( Complex{Float64}(k0^2 - Kx^2 - Ky^2) )) for Kx in coords.kX, Ky in coords.kY]
```
which leads to **gain** for negative Im(eps) due to complex conjugation _after_  multiplying k with sqrt(eps)
Removing the complex conjugation here seems the obvious solution but I would like to understand why it's there in the first place.
If Kx^2 > k0^2, sqrt() would lead to a positive imaginary part meaning gain, I guess that's why it was conjugated? I don't quite understand this part (see the issue I opened a while back). 

Alternatively we set 
`k0 = 2*pi/lambda*conj(sqrt(eps))` 

which seems to work but this getting a bit confusing with so many complex conjugations...

jhbdiehl commented:
So. Sorry if I ignored a previous issue.

Your point is completely valid! Good catch! I'm pretty positive that the reason for the `conj()` is exactly what you say it is.

There is a way not to require the `conj()` (and the explizit specification of `Complex{Float64}`):

`k_prop = [sqrt(k0^2 - Kx^2 - Ky^2 - 0.0im)  for Kx in coords.kX, Ky in coords.kY]`

This always gives the negative root if the ks are all real. If k0 is not real this will not change the sign of the imaginary part. So an Im(eps) will translate to an Im(k_prop). So this is exactly what we want, right?

Edit: I guess I won't push that since it's not an mppmu branch. But please have a look, change the line and maybe write a comment if you think that's sensible

J-co commented:
Seems reasonable for our purposes. Not sure if it is really correct, there is some interaction with complex permitivity and cut off frequency that play into here I think. I'll have to dig more into the literature...

Maybe this is also a good point to start thinking about more sophisticated loss models...oh well..

J-co committed:
Update boost_fractor.jl

Made sure k_prop is consistent across different propagators

J-co commented:
According to this paper https://hal-centralesupelec.archives-ouvertes.fr/hal-01589787/document
the interpretation of plane wave expansion in a lossy medium needs a fourier transform with complex Kx,Ky (aka laplace transformation) to still make sense. However, if I understand correctly, as long as one just concerned with propagation, i.e. is always transforming back into real space one can still use real Kx, Ky. 
So I changed to use Johannes' suggestion and also made the 1D propagator look more consistent with the rest...

jhbdiehl commented:
Hmm. Checks seem to be failing, which is bothersome for such a tiny change.
Do you look into it or should I?

J-co commented:
Seems like it changes the boostfactor? I'll look into it...Do checks test for complex epsilon?

J-co committed:
Tracking down failing tests...

jhbdiehl commented:
All tests use epsilon = 9, so no. Strange. But yes, looking into the logs the code is still running, but it produces a bf that's different...

J-co committed:
Revert back to original

J-co commented:
Ok, so it passes for the original

```
k0 = 2*pi/lambda*sqrt(eps)
k_prop = [conj(sqrt( Complex{Float64}(k0^2 - Kx^2 - Ky^2) )) for Kx in coords.kX, Ky in coords.kY]
```
which is wrong since it leads to gain for negative Im(eps) and it fails for 

```
k0 = 2*pi/lambda*conj(sqrt(eps))
k_prop = [conj(sqrt( Complex{Float64}(k0^2 - Kx^2 - Ky^2) )) for Kx in coords.kX, Ky in coords.kY]
```
which would be correct. Same thing for your cleaner suggestion, Johannes.
I tested it myself, even with real epsilons, this makes a (small) difference! Not sure what is going on, maybe numerical precision but why would it change if conj(sqrt(eps))===sqrt(eps) for real eps...

jhbdiehl commented:
Yeah, I'm looking into it too right now. Super confused...
Just changing one line it passes for
`
k_prop = [conj(sqrt( Complex{Float64}(k0^2 - Kx^2 - Ky^2) )) for Kx in coords.kX, Ky in coords.kY]
`
Doesn't pass for
`
k_prop2 = [sqrt(k0^2 - Kx^2 - Ky^2-0.0im)  for Kx in coords.kX, Ky in coords.kY]
`
but `k_prop == k_prop2` returns `true`.

J-co commented:
My best guess is, that since eps is always casted to be complex in the SetupBoundaries struct, it somehow makes a difference whether it has +-0.0im (we know it does for sqrt()) and then this propagates as a numerical error to the boostfactor...

J-co commented:
While 
```

1.0-0.0im ==1.0+0.0im => true
1.0-0.0im ===1.0+0.0im => false
```
So I guess` k_prop === k_prop2 => false`

jhbdiehl commented:
Yes.
`
a = 1.0 + 0.0im
`
`
a - 0.0im
`
returns `1.0 + 0.0im.`

But
`
a = 1.0 - 0.0im
`
`
a + 0.0im
`
also returns `1.0 + 0.0im.`

Therefore taking -0.0im was a bad idea since it doesn't do anything if eps is already specified as a complex number.

And, importantly the difference in the bf is not only numerical, since for the given tests a part of the k_prop array becomes imaginary and the other solution of the sqrt is chosen. Is that a problem?

jhbdiehl commented:
Ok, so I have a solution but it ain't pretty:

Adding

`imag(eps) == 0.0 ? eps = real(eps) - 0.0im : nothing`

should solve all our problems. This line makes sure that if eps is real the imaginary part is negative instead of positive. This doesn't change anything if `k_prop` is real, but if it is imaginary it makes sure that the negative root is chosen and we actually have loss and not gains. So in this regard it does the same as `conj()` did before.

If eps is complex this line won't do anything, so it will not wrongly make gains out of losses like `conj()` did.

My implementation survives the runtests, let me know what you think!

Another solution would be to make a distinction in the sbrdy object between real and complex epsilon. Maybe is more elegant but I dont know how to do it with less code.



For completeness:
`
k0 = 2*pi/lambda*sqrt(eps)
k_prop0 = [sqrt(k0^2 - Kx^2 - Ky^2)  for Kx in coords.kX, Ky in coords.kY]
`

J-co commented:
We could add
`imag(eps) == 0.0 ? eps = real(eps) - 0.0im : nothing`
in the constructor of SetupBoundaries which isn't too ugly. I guess we can think of it as zero loss (instead of zero gain as the vanilla +0.0im casting from real to complex would imply) which I think is a good way to rationalize it...
I'll do some further testing tomorrow, if anything this showed that there are still things to be scrutinized...

J-co committed:
Update boost_fractor.jl

Made sure that real epsilon are casted as real(eps) - 0.0im in SeedSetupBoundaries instead of real(eps) + 0.0im

J-co committed:
Small fix

J-co committed:
Update boost_fractor.jl

Removed unnecessary second cast to Complex{Float64}

J-co commented:
Ok, I opted for adding 
```
epsilon = Array{Complex{Float64}}(epsilon)
epsilon[imag.(epsilon) .== 0.0] .= real.(epsilon[imag.(epsilon) .== 0.0]) .- 0.0im
```
in the SeedSetupBoundaries constructor. This way real epsilons are always represented as Re(eps) - 0.0im. 

I wonder if we should put an additional check in the propagator function if someone is using it manually?

jhbdiehl commented:
I mean, if we do, we could have just implemented the line there, couldn't we?
But at least add a comment stating the problem and why the line is there, then I'm happy! If you want to do more, fine.

J-co commented:
True, but it doesn't seem right to have the "wrong" epsilon in the sbdry struct and then change it everytime in the propagator.
I'll add the check, just to make sure. Also a comment...

J-co committed:
Update boost_fractor.jl

Added additional check in propagator just to make sure Im(eps) = -0.0 for real eps

jhbdiehl reviewed (approved):
Looks good! I'll merge.
